enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,634 @@
+// IngressControllerSetOrDeleteHTTPHeader specifies options for set/delete a header.
+type IngressControllerSetOrDeleteHTTPHeader struct {
+    // name specifies a header name to be set(replace)/append/delete.  Its value must be a valid HTTP header.
+    // name as defined in RFC 2616 section 4.2.
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=256
+    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+    // +required
+    Name string `json:"name"`
+
+    IngressControllerSetOrDeleteHTTPHeaderUnion `json:",inline"`
+}
+```
+
+```go
+// IngressControllerSetOrDeleteHTTPHeaderUnion specifies fields to mention the header details and the action you want to perform on them.
+// +union
+type IngressControllerSetOrDeleteHTTPHeaderUnion struct {
+	// action lets you specify headers for Set or Delete.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:=Set;Delete
+	// +kubebuilder:validation:Required
+	Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`
+
+	// Set defines HTTP headers that should be set.
+	// If this field is empty, no headers are added.
+	//
+	// Note that this option only applies to cleartext HTTP connections
+	// and to secure HTTP connections for which the ingress controller
+	// terminates encryption (that is, edge-terminated or reencrypt
+	// connections).  Headers cannot be set(replace)/append for TLS passthrough
+	// connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
+
+	// delete defines HTTP headers that should be deleted.
+	// If this field is empty, no headers are deleted.
+	//
+	// Note that this option only applies to cleartext HTTP connections
+	// and to secure HTTP connections for which the ingress controller
+	// terminates encryption (that is, edge-terminated or reencrypt
+	// connections).  Headers cannot be set(replace)/append for TLS passthrough
+	// connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`
+}
+```
+
+```go
+// IngressControllerSetAction defines operation you can perform on headers.
+// +kubebuilder:validation:Enum=Set;Delete
+type IngressControllerSetHTTPHeaderAction string
+
+const (
+    Set    IngressControllerSetHTTPHeaderAction = "Set"
+    Delete IngressControllerSetHTTPHeaderAction = "Delete"
+)
+```
+
+The IngressControllerSetHTTPHeader type has fields for specifying a header value.
+```go
+// IngressControllerSetHTTPHeader what value of HTTP header needs to be set.
+type IngressControllerSetHTTPHeader struct {
+    // Value specifies a header value.
+    // +kubebuilder:validation:Required
+    // +required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=4096
+    Value string `json:"value"`
+}
+```
+
+The IngressControllerDeleteHTTPHeader type is an empty union member .
```

Comment:
Hm, I guess "empty union member" means that there is no member field.  I have been using the phrase "empty union member" when the field exists but its type is an empty struct.  In that case, "a union member that is an empty struct" would be technically more correct (but it does seem a bit verbose).

Reply:
Ah ok, now I see what you meant by "empty union member". Yeah, that has a different definition in the conventions doc. I think it's better to avoid using "empty union member" for Delete as it may be misleading for others too. We can rephrase it somehow or remove.